app/utils/actions_for_assistants/actions_for_assistants.py

```diff
@@ -21,117 +31,251 @@ def read_text_file(file_path):
+    })
+    return result.content
+
+def rewrite_tool(user_query: str, focus: str, tone: str):
+    """ Used for rewriting content"""
+    prompt = ChatPromptTemplate.from_messages(
+        [
+            (
+                "system",
+                read_text_file("prompts/system/rewrite_tool.txt")
+            ),
+            (
+                "human",
+                read_text_file("prompts/human/rewrite_tool.txt")
+            )
+        ]
+    )
+
+    chain = prompt | chat_google_genai
+    result = chain.invoke({
+        "user_query": user_query,
+        "focus": focus,
+        "tone": tone
+    })
+    return result.content
+
+def custom_prompt_tool(user_query: str, action: str):
+    """ Used for managing custom prompts"""
     prompt = ChatPromptTemplate.from_messages(
+        [
+            (
+                "system",
+                read_text_file("prompts/system/custom_prompt_tool.txt")
+            ),
+            (
+                "human",
+                read_text_file("prompts/human/custom_prompt_tool.txt")
+            )
+        ]
+    )
+
+    chain = prompt | chat_google_genai
+    result = chain.invoke({
+        "user_query": user_query,
+        "action": action
+    })
+    return result.content
+
+
+def generate_questions_to_json_tool(user_query: str):
+    """ Used for generating questions. The results are returned in a JSON format. You must respect the user's language """
+    prompt_generate_questions = ChatPromptTemplate.from_messages(
+      [
+          (
+            "system",
+            read_text_file("prompts/system/question_generation.txt")
+          ),
+          ("human",
+            read_text_file("prompts/human/question_generation.txt")
+          )
+      ]
+    )
+
+    chain_generate_questions = prompt_generate_questions | chat_google_genai
+
+    parser = JsonOutputParser(pydantic_object=QuestionList)
+    prompt_questions_to_json = ChatPromptTemplate.from_messages(
     [
         (
             "system",
-            read_text_file('prompts/system_prompt_question_generation_json.txt')
+            read_text_file("prompts/system/generate_questions_to_json_tool.txt")
         ),
-        ("human", read_text_file('prompts/user_prompt_question_generation_json.txt'))
+        ("human",
+            read_text_file("prompts/human/generate_questions_to_json_tool.txt")
+        )
     ]
     )
 
-    chain = prompt | chat_google_genai
+    chain_questions_to_json = prompt_questions_to_json | chat_google_genai
+    combined_chain = (
+        chain_generate_questions
+        | RunnableLambda(lambda x: adapt_questions_for_json(x.content, parser))
+        | chain_questions_to_json
+        | parser
+    )
+
+    result = combined_chain.invoke(user_query)
 
-    result = chain.invoke(
-        {
-            "questions": state["result"],
-            "format_instructions": parser.get_format_instructions()
-        }
+    return result
+
+def call_marvel_ai_tool(user_query: str):
```

Comment:
Good POC but we need to investigate this a bit more and come up with a more declarative way where our LLM can make better interpretation of what tool to call, just like it does for other actions like summarize etc. Lets omit this feature in this PR

Reply:
Done